### ISSUE TITLE
Desire for more detailed readme documentation

### ISSUE THREAD

**Issue by thanksforyourworks** [annegnx/PnP-Flow#2]
Thank you for doing a really great job in the field of image denoising, I am trying to learn your code. However, the Readme doesn't provide a description of the folders and files in the project, which makes me wonder what the code in the “ImageGeneration” folder does and how to run it. I also noticed that the denoising network used in the project for training the denoising task is: src/train_denoiser.py, which seems to be different from the method used in the paper and the demo. I would appreciate if you could add some textual clarification to help me understand.
Finally, thank you for your time. I'm just a beginner and maybe the questions aren't appropriate. I wish you all the best in your endeavors.

**Comment by annegnx** [comment]
Thank you for your interest!
The code in Image Generation is copied from [this repo](https://github.com/gnobitab/RectifiedFlow) so as to facilitate the use of Rectified pre-trained models on the different implemented method. This reuse is mentioned in the Acknowledgments section of the README.

This repo can be used to benchmark different methods/models, this is why there are two different `train_*.py` files. `train_denoiser.py` trains a model to remove noise from image (it is *not* a generative model), while the other file trains flow matching models.  Our method (PnP-Flow) indeed works with Flow Matching, the `train_denoiser.py` is used in the paper to benchmark against PnP-GS.

Thanks again for your feed-back, we will enrich the documentation for better clarity.

**Comment by thanksforyourworks** [comment]
Thank you for your time, I was pleasantly surprised to receive your reply. Allow me to ask you some more questions to learn more about your work.
First of all, I'd like to thank you very much for taking my feedback, and I'll always look forward to your updates to the docs!
Secondly, I would like to confirm my understanding of your answer to make sure I have not misunderstood: 1. Image Generation exists to use the pre-trained models in Rectified Flow. 2. train_denoiser.py is used as a baseline model comparison. 3. If I wish to use your excellent work to train the denoising on other datasets task, I should use flow matching to learn the stylistic features of the dataset and then use the (aka plug-and-play) denoiser in the demo, combined with the pre-trained flow matching and interpolation methods to accomplish it.
Once again, I sincerely apologize for the additional workload I've put on you, and I hope you'll forgive a beginner's lack of knowledge and poor comprehension. I'm beyond looking forward to your additional replies, thanks!

**Comment by thanksforyourworks** [comment]
Thank you for your time, I was pleasantly surprised to receive your reply. Allow me to ask you some more questions to learn more about your work.
First of all, I'd like to thank you very much for taking my feedback, and I'll always look forward to your updates to the docs!
Secondly, I would like to confirm my understanding of your answer to make sure I have not misunderstood: 1. Image Generation exists to use the pre-trained models and utils in Rectified Flow. The actual denoiser is used like the demo. 2. train_denoiser.py is used as a baseline model comparison. 3. If I wish to use your excellent work to train the denoising on other datasets task, I should use flow matching to learn the stylistic features of the dataset and then use the (aka plug-and-play) denoiser in the demo, combined with the pre-trained flow matching and interpolation methods to accomplish it. Additionally, the hyperlink in the second paragraph of your last reply doesn't seem to be showing up, so I hope you'll add it if it's convenient for you.
Once again, I sincerely apologize for the additional workload I've put on you, and I hope you'll forgive a beginner's lack of knowledge and poor comprehension. I'm beyond looking forward to your additional replies, thanks!

**Comment by thanksforyourworks** [comment]
Thank you for your time, I was pleasantly surprised to receive your reply. Allow me to ask you some more questions to learn more about your work.
First of all, I'd like to thank you very much for taking my feedback, and I'll always look forward to your updates to the docs!
Secondly, I would like to confirm my understanding of your answer to make sure I have not misunderstood: 1. Image Generation exists to use the pre-trained models and utils in Rectified Flow. The actual denoiser is used like the demo. 2. train_denoiser.py is used as a baseline model comparison. 3. If I wish to use your excellent work to train the denoising on other datasets task, I should use flow matching to learn the stylistic features of the dataset and then use the (aka plug-and-play) denoiser in the demo, combined with the pre-trained flow matching and interpolation methods to accomplish it. 4、Where will the project src/methods/pnp_flow.py be used? Additionally, the hyperlink in the second paragraph of your last reply doesn't seem to be showing up, so I hope you'll add it if it's convenient for you.
Once again, I sincerely apologize for the additional workload I've put on you, and I hope you'll forgive a beginner's lack of knowledge and poor comprehension. I'm beyond looking forward to your additional replies, thanks!

**Comment by annegnx** [comment]
1. Rectified Flow is one type of Flow Matching model (that can be used instead of OT Flow matching for example). Our method relies on pre-trained Flow Matching models: we provide a script `src/train_flow_matching.py` to train your own Flow Matching model or you can use pre-trained models (e.g. checkpoints available on the Rectified Flow repo).  
2. Yes, not all image restorations methods rely on generative models. Our repo also allows to benchmark against methods which use denoising models instead. 
To use our PnP-Flow method on your new dataset: 
- first, you need to add a class for your dataset in the `src/dataloaders.py` file
- add a config file for your dataset (in `config/dataset_config`)
-  train a FM model (using `src/train_flow_matching.py` -> look at `scripts/script_train.sh` and replace the dataset with your dataset) 
-  then, you can use our method (which is implemented in `src/methods/pnp_flow.py` ) ->  look at `scripts/script_test.sh` where you can specify your dataset and choose `model=pnp_flow` to implement our method.

**Comment by thanksforyourworks** [comment]
Thank you for your detailed and patient response, I will continue to learn from your work based on your guidance. I will seek your guidance again if I have any other questions. Thanks again for your additional work!

**Comment by annegnx** [comment]
I just closed the issue, but no to hesitate to re-open it if you have other questions !

**Comment by thanksforyourworks** [comment]
Thank you very much for your patience and care!&nbsp;If I have any questions, I will reopen the issue and you can answer at your convenience, even if it's a week later. &nbsp;In fact, very few scholars can reply in time to the problems that arise on GitHub, which is also the reason why I am surprised. Your generous guidance has been very helpful to me, and I wish you all the best!
------------------ Original ------------------
From: annegnx ***@***.***&gt;
Date: Fri,Dec 6,2024 5:39 PM
To: annegnx/PnP-Flow ***@***.***&gt;
Cc: thanksforyourworks ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [annegnx/PnP-Flow] Desire for more detailed readme documentation(Issue #2)





 
I just closed the issue, but no to hesitate to re-open it if you have other questions !

**Comment by thanksforyourworks** [comment]
I'm sorry to bother you again, I'm having some problems running the program. Would it be convenient for you to provide me with the version of python you are experimenting with and the corresponding specific version in the requirements.txt file. I would appreciate it if you could reply again!

**Comment by wrqcodedoge** [comment]
The following environment works for me, with Python 3.10.0.
Package                  Version
------------------------ -----------
absl-py                  2.2.0
aiohappyeyeballs         2.6.1
aiohttp                  3.11.14
aiosignal                1.3.2
annotated-types          0.7.0
async-timeout            5.0.1
attrs                    25.3.0
autograd                 1.6.2
certifi                  2025.1.31
charset-normalizer       3.4.1
click                    8.1.8
contourpy                1.3.1
cycler                   0.12.1
Cython                   3.0.10
deepinv                  0.2.0
diffeqtorch              1.0.0
docker-pycreds           0.4.0
einops                   0.8.1
filelock                 3.18.0
fonttools                4.56.0
frozenlist               1.5.0
fsspec                   2025.3.0
future                   1.0.0
gitdb                    4.0.12
GitPython                3.1.44
gmpy2                    2.2.1
grpcio                   1.71.0
h5py                     3.13.0
hdf5storage              0.1.19
idna                     3.10
imageio                  2.37.0
jax                      0.5.3
jaxlib                   0.5.3
Jinja2                   3.1.6
joblib                   1.4.2
julia                    0.6.2
keopscore                2.2.3
kiwisolver               1.4.8
lazy_loader              0.4
lightning                2.5.1
lightning-utilities      0.14.2
lpips                    0.1.4
Markdown                 3.7
MarkupSafe               3.0.2
matplotlib               3.10.1
ml_collections           1.0.0
ml_dtypes                0.5.1
mpmath                   1.3.0
multidict                6.2.0
networkx                 3.4.2
ninja                    1.11.1.3
numdifftools             0.9.41
numpy                    2.2.4
nvidia-cublas-cu12       12.4.5.8
nvidia-cuda-cupti-cu12   12.4.127
nvidia-cuda-nvrtc-cu12   12.4.127
nvidia-cuda-runtime-cu12 12.4.127
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.2.1.3
nvidia-curand-cu12       10.3.5.147
nvidia-cusolver-cu12     11.6.1.9
nvidia-cusparse-cu12     12.3.1.170
nvidia-cusparselt-cu12   0.6.2
nvidia-nccl-cu12         2.21.5
nvidia-nvjitlink-cu12    12.4.127
nvidia-nvtx-cu12         12.4.127
opencv-python            4.11.0.86
opt_einsum               3.4.0
packaging                24.2
pandas                   2.2.3
pillow                   11.1.0
pip                      24.2
platformdirs             4.3.7
POT                      0.9.5
propcache                0.3.0
protobuf                 5.29.4
psutil                   7.0.0
pybind11                 2.13.6
pydantic                 2.10.6
pydantic_core            2.27.2
pykeops                  2.2.3
pyparsing                3.2.1
python-dateutil          2.9.0.post0
pytorch-ignite           0.5.1
pytorch-lightning        2.5.1
pytz                     2025.1
PyYAML                   6.0.2
requests                 2.32.3
scikit-image             0.25.2
scikit-learn             1.6.1
scipy                    1.15.2
sentry-sdk               2.24.0
setproctitle             1.3.5
setuptools               75.1.0
six                      1.17.0
smmap                    5.0.2
sympy                    1.13.1
tensorboard              2.19.0
tensorboard-data-server  0.7.2
threadpoolctl            3.6.0
tifffile                 2025.3.13
toolz                    1.0.0
torch                    2.6.0
torchdiffeq              0.2.5
torchmetrics             1.7.0
torchvision              0.21.0
tqdm                     4.67.1
triton                   3.2.0
typing_extensions        4.12.2
tzdata                   2025.1
urllib3                  2.3.0
wandb                    0.19.8
Werkzeug                 3.1.3
wheel                    0.44.0
yarl                     1.18.3
